tech-specs/src/explore/design/dkg.md

```diff
@@ -1 +1,157 @@
+message is handled by the Ferveo library.
+
+The signed message will be returned as part of the vote extension and there
+is a special verification that must be done, otherwise the PVSS transcript 
+should not be included in the aggregation (discussed next). This verification
+is provided by the Ferveo library. The PVSS transcripts will be included in
+the header of the finalized block along with the aggregation after the verify
+vote extension phase completes. This is necessary so that validators can
+verify the aggregation.
+
+## Aggregation
+
+At the end of the vote extension phase, enough validators 
+(meaning their combined weight  shares exceeds \\( \frac{2}{3}\\) of the 
+total) *should* have shared their PVSS to the current proposer and hopefully
+enough of these transcripts are valid. This is because the vote extension 
+phase requires at least \\( \frac{2}{3}\\) of validators to send a vote
+extension before it can move to the finalization phase.
+
+At this point, the next block proposer can aggregate the messages into a 
+single instance using Ferveo as part of the verify vote extension phase. 
+This is then included in the header of the finalized block along with the
+PVSS transcripts. Once this DKG is on the blockchain, it can be 
+used for encryption of transactions until the next DKG instance is run.
+
+When deciding to aggregate the PVSS instances into a single one, we shall 
+consider the instances ranked in terms of their validators as was also used
+for partitioning the weight shares (described above).
+
+## Storage considerations
+
+The storage of a node must be adapted to accommodate the DKG protocol. Each
+instance of a DKG protocol from Ferveo is a state machine that stores
+much of the necessary data from the protocol instance. The following data
+needs to be stored by each node
+ - The partition of the weight shares among the current validator set
+ - A session keypair for the current DKG session. Used for encrypting
+   PVSS transcripts and signing DKG protocol messages.
+ - A list of session ids from the last completed DKG instance up until the
+   latest started (we may prune ids for failed instances however) 
+ - The DKG state machine associated to each of these session ids.
+
+##  The DKG state machine
+
+We first describe each new message type needed for the protocol.
+- `Deal`: This message contains the PVSS transcript from a validator
+  acting as a dealer
+- `Aggregate`: Combine the PVSS instances into a single one, thereby
+  producing the final key.
+- `Complain`: A message stating that a given validator's PVSS transcript or
+  a block proposer's aggregation failed verification. 
+
+In Ferveo, the DKG protocol is modelled as a state machine. This allows a
+validator to initialize an instance and drive it forward by reacting to
+each of the above message in the appropriate way. All necessary state
+is persisted by this state machine. We describe the  appropriate action for
+each of the above transactions.
+
```

Comment:
I think I understand this a bit differently:
Each validator sends a `Deal` message with their PVSS. The DKG listens to these messages until a threshold has been met (66%). Then it (internally) triggers the shares generation step. Once shares have been distributed, a node sends an `Aggregate` message which is added to the block header, or a `Complain` if it can't aggregate the transcripts succesfully, which in turn sets the DKG state machine to an Invalid state.
After this `Aggregate` message, we still need consensus from validators that `Check` this aggregation. They can also issue a `Complain` message if they don't succeed.

Reply:
<img width="537" alt="Screenshot 2021-08-19 at 11 41 23" src="https://user-images.githubusercontent.com/2612856/130055423-48e5f240-23b0-4464-9337-6347dd8c4514.png">

Reply:
We could do that, but I think it would be better to aggregate in as many validators as possible. That way we don't limit which 2/3 set can be used later for decryption (in case of validators going down etc).

Reply:
This diagram just indicates that at least 2/3 should deal, not that we only aggregate the fist 2/3 that are received. I'm more in favor of  aggregatoing all shares accrued with the `n` blocks of time designated. If that is 99% of the validators, that is even better right? It should just be at least 2/3

Reply:
Actually, re-looking at how vote extensions work, this may indeed be how we want to do it.

Reply:
Actually, I'm quickly becoming convinced. @joebebel What do you think? The big question is how to accomodate spreading of sharing PVSS transcripts across multiple blocks.

Reply:
Ok, after much confusion on my part, I understand what you're saying. I will make the appropriate changes.

Reply:
We'd also need to discuss what the contents of the `Invalid` state are, so that the next block proposer can lookup what went wrong. E.g. the `complain` message from an invalid pvss during aggregation may contain different data than the `complain` message from finding the aggregation itself invalid

Reply:
This is a more conventional graph where messages are ellipses and blue rectangles are states. Hope it helps. I'm still not clear with the complain messages

<img width="543" alt="Screenshot 2021-08-19 at 15 27 38" src="https://user-images.githubusercontent.com/2612856/130086836-dce755d6-71bc-4444-be49-fe84ed9ebc0e.png">

Reply:
Right, I think we still have to figure out how the Invalid state can recover back to to the Deal/Aggregate states (mostly just pretend the bad data doesn't exist, I guess?)